### PR TITLE
Enum.take_random/2: Correct typespec, guards and optimize

### PR DESCRIPTION
**Pull request by eksperimental**

### CODE REVIEW

File: lib/elixir/lib/enum.ex [hunk 1]
```diff
@@ -2247,18 +2247,21 @@ defmodule Enum do
...
-  def take_random(first..last, 1) when first > last do
+  def take_random(first..first, 1) when is_integer(first),
+    do: [first]
+
+  def take_random(first..last, 1) when is_integer(first) and is_integer(last) and first > last do
     take_random(last..first, 1)
   end
 
-  def take_random(first..last, 1) do
+  def take_random(first..last, 1) when is_integer(first) and is_integer(last) do
```
Comment (lexmag) [inline lib/elixir/lib/enum.ex]: That's range guarantee that both sides are integers.

Comment (lexmag) [inline lib/elixir/lib/enum.ex]: Same applies to all `is_integer/1` check above.

Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: Technically u can create a non valid range via `%Range{}` that will match `_.._` and that wont be valid

Comment (whatyouhide) [inline lib/elixir/lib/enum.ex]: I would say that the `%Range{}` API is not public API? I mean it's visible in the documentation for `Range` (as `Range.t`) but still.

Comment (lexmag) [inline lib/elixir/lib/enum.ex]: Technically we can do many crazy things, e.g. send `:DOWN` message to a process. There are many places where we expect only valid data (`MapSet`, `Stream`, etc.).

**Comment by eksperimental** [comment]
Im all in for removing this overhead, but if i recall correctly it was mentioned somewhere about creating ranges with this api. Elixir lang org, or the docs. Im not with the computer right now.. will check it later. Cant recall wether it was about creating or matching ranges this was

File: lib/elixir/lib/enum.ex (hunk 1 above)
Comment (josevalim) [inline lib/elixir/lib/enum.ex]: Yeah, the integer checks for range should not be necessary IMO.

**Comment by eksperimental** [comment]
I wasn't able to find any reference to this API, so it was probably removed, or I read it somewhere else.
I have added a second commit removing this unncessary check

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**